REGR: fix bug in DatetimeIndex slicing with irregular or unsorted indices

CloseChoice opened the pull request:
closes #36953 
closes #35509
- [x] tests added / passed
- [ ] passes `black pandas`
- [ ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

CloseChoice committed:
remove assert in core/generic.py and add test for datetimeindex slicing

pep8speaks commented:
Hello @CloseChoice! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:



* In the file [`pandas/tests/generic/test_generic.py`](https://github.com/pandas-dev/pandas/blob/f1d4a102c24919bf4b467b5f1ce96d8f080c61e3/pandas/tests/generic/test_generic.py):

> [Line 903:89](https://github.com/pandas-dev/pandas/blob/f1d4a102c24919bf4b467b5f1ce96d8f080c61e3/pandas/tests/generic/test_generic.py#L903): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (91 > 88 characters)
> [Line 908:1](https://github.com/pandas-dev/pandas/blob/f1d4a102c24919bf4b467b5f1ce96d8f080c61e3/pandas/tests/generic/test_generic.py#L908): [W391](https://duckduckgo.com/?q=pep8%20W391) blank line at end of file

CloseChoice committed:
fix formatting in test_generic

arw2019 commented on pandas/core/generic.py:
```diff
@@ -3736,7 +3736,7 @@ def _slice(self: FrameOrSeries, slobj: slice, axis=0) -> FrameOrSeries:
         Slicing with this method is *always* positional.
         """
-        assert isinstance(slobj, slice), type(slobj)
+        # assert isinstance(slobj, slice), type(slobj)
```
just remove this

CloseChoice replied:
done

arw2019 commented on pandas/tests/generic/test_generic.py:
```diff
@@ -897,3 +897,13 @@ def test_flags_identity(self, as_frame):
         s2 = s.copy()
         assert s2.flags is not s.flags
+
+    def test_slice_incomplete_datetimes(self):
```
I'd call it `test_slice_irregular_datetime_index_with_nan`

CloseChoice replied:
done.

arw2019 commented on pandas/tests/generic/test_generic.py:
```diff
@@ -897,3 +897,13 @@ def test_flags_identity(self, as_frame):
+
+    def test_slice_incomplete_datetimes(self):
+        # GH36953 and GH35509
+        increment = pd.Series(
```
I know this is the OP but is it possible to explicitly construct it? without using the `cumsum`

CloseChoice replied:
done

arw2019 commented on pandas/tests/generic/test_generic.py:
```diff
@@ -897,3 +897,13 @@ def test_flags_identity(self, as_frame):
+        increment[-1] = None
+        index = pd.Timestamp("2012-01-01 09:00") + increment
+        df = pd.Series(range(len(index)), index=index).to_frame()
+        assert len(df["2012-01-01":"2012-01-05"]) == 1000
```
can you construct the expected result and compare against that with `tm.assert_frame_equal`?

CloseChoice replied:
done

CloseChoice committed:
fix test and remove commented line

arw2019 commented on pandas/tests/generic/test_generic.py:
```diff
@@ -897,3 +897,13 @@ def test_flags_identity(self, as_frame):
+        ).cumsum()
+        increment[-1] = None
+        index = pd.Timestamp("2012-01-01 09:00") + increment
+        df = pd.Series(range(len(index)), index=index).to_frame()
```
could use `df = pd.DataFrame(range(len(index)), index=index)`

CloseChoice replied:
done

arw2019 reviewed (changes requested):
Thanks @CloseChoice for the PR!

Some comments.  

Can you add the OP from #35509? As per https://github.com/pandas-dev/pandas/issues/36953#issuecomment-706220412 the two issues are caused by the same underlying bug but different behaviors so we'd like to have tests for both

I'm also not sure about the current location for this. Haven't looked in detail but somewhere like `pandas/tests/frame/test_timeseries.py` might be more appropriate

CloseChoice committed:
updated tests according to PR comments

CloseChoice commented:
> Thanks @CloseChoice for the PR!
> 
> Some comments.
> 
> Can you add the OP from #35509? As per [#36953 (comment)](https://github.com/pandas-dev/pandas/issues/36953#issuecomment-706220412) the two issues are caused by the same underlying bug but different behaviors so we'd like to have tests for both
> 
> I'm also not sure about the current location for this. Haven't looked in detail but somewhere like `pandas/tests/frame/test_timeseries.py` might be more appropriate

Hi @arw2019 thanks for the quick review. I added another test corresponding to #35509 and moved both tests to your suggested destination `pandas/tests/frame/test_timeseries.py`.

CloseChoice committed:
slice with loc to prevent future warning

CloseChoice committed:
revert changes in io/formats/css.py

CloseChoice committed:
run black on test_timeseries.py

simonjayhawkins reviewed (changes requested):
Thanks @CloseChoice can you add a release note to doc/source/whatsnew/v1.1.4.rst

jreback commented on pandas/tests/frame/test_timeseries.py:
```diff
@@ -63,3 +63,22 @@ def test_datetime_assignment_with_NaT_and_diff_time_units(self):
         )
         tm.assert_frame_equal(result, expected)
+
+    def test_slice_irregular_datetime_index_with_nan(self):
```
move these to pandas/tests/indexing/test_partial.py

*always* use
result=
expected=
tm.assert_.....

jreback commented:
also this needs a whatsnew note for 1.1.4 in regressions (mention both issues)

CloseChoice committed:
move tests to indexing/test_partial, add whatsnew entry

CloseChoice committed:
reformat test_timeseries.py

CloseChoice committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into BUG-regression-datetimeindex-slicing

CloseChoice committed:
update whatsnew

mroeschke commented on pandas/tests/indexing/test_partial.py:
```diff
@@ -681,3 +681,22 @@ def test_index_name_empty(self):
+        index = pd.to_datetime(["2012-01-01", "2012-01-02", "2012-01-03", None])
+        df = pd.DataFrame(range(len(index)), index=index)
+        expected = pd.DataFrame(range(len(index[:3])), index=index[:3])
+        tm.assert_frame_equal(df["2012-01-01":"2012-01-04"], expected)
```
Could you define `result = df["2012-01-01":"2012-01-04"]` above?

mroeschke commented on pandas/tests/indexing/test_partial.py:
```diff
@@ -681,3 +681,22 @@ def test_index_name_empty(self):
+            {"col1": ["a", "c"], "col2": [1, 3]},
+            index=pd.to_datetime(["2020-08-01", "2020-08-05"]),
+        )
+        tm.assert_frame_equal(df.loc["2020-08"], expected)
```
Could you define `result = df["2020-08"]` above?

jreback commented on pandas/core/generic.py:
```diff
@@ -3739,7 +3739,6 @@ def _slice(self: FrameOrSeries, slobj: slice, axis=0) -> FrameOrSeries:
 
         Slicing with this method is *always* positional.
         """
-        assert isinstance(slobj, slice), type(slobj)
```
this is not the patch; something is going down the wrong path and ending up here. this routine is *only* for slices.

CloseChoice replied:
Ok, I will look into that. I just wanted to note that in version 1.0.9 it went down just the same path except that the exception wasn't there yet. In this sense we are not talking about a regression but about fixing a bug which seems to have been there since a while ago.

jorisvandenbossche replied:
@CloseChoice can you check if the suggestion from here https://github.com/pandas-dev/pandas/issues/35509#issuecomment-687430169 passes the tests you added?

CloseChoice committed:
Merge branch 'master' into BUG-regression-datetimeindex-slicing

CloseChoice committed:
add suggestion from comment

CloseChoice committed:
use result and expected in tests

CloseChoice committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into BUG-regression-datetimeindex-slicing

simonjayhawkins commented on doc/source/whatsnew/v1.1.4.rst:
```diff
@@ -22,6 +22,7 @@ Fixed regressions
 - Fixed regression in :class:`RollingGroupby` causing a segmentation fault with Index of dtype object (:issue:`36727`)
 - Fixed regression in :meth:`DataFrame.resample(...).apply(...)` raised ``AttributeError`` when input was a :class:`DataFrame` and only a :class:`Series` was evaluated (:issue:`36951`)
 - Fixed regression in :class:`PeriodDtype` comparing both equal and unequal to its string representation (:issue:`37265`)
+- Fixed regression where slicing :class:`DatetimeIndex` failed on irregular time series with ``pd.NaT`` or on unsorted indices (:issue:`36953` and :issue:`35509`)
```
```suggestion
- Fixed regression where slicing :class:`DatetimeIndex` raised :exc:`AssertionError` on irregular time series with ``pd.NaT`` or on unsorted indices (:issue:`36953` and :issue:`35509`)
```

simonjayhawkins reviewed (approved):
Thanks @CloseChoice lgtm pending green. are failures unrelated to the changes in this PR?

CloseChoice committed:
Update doc/source/whatsnew/v1.1.4.rst

Co-authored-by: Simon Hawkins <simonjayhawkins@gmail.com>

CloseChoice commented:
> Thanks @CloseChoice lgtm pending green. are failures unrelated to the changes in this PR?

I don't see any connection between this PR and I see some of the failing tests in other PRs aswell. But might be worth an investigation.

jreback commented:
k this looks fine, but I think there are some failing tests that are related: https://travis-ci.org/github/pandas-dev/pandas/jobs/737504953

so pls merge master and let's see what the CI says

CloseChoice committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into BUG-regression-datetimeindex-slicing

CloseChoice committed:
Merge branch 'BUG-regression-datetimeindex-slicing' of github.com:CloseChoice/pandas into BUG-regression-datetimeindex-slicing

CloseChoice committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into BUG-regression-datetimeindex-slicing

simonjayhawkins committed:
Merge remote-tracking branch 'upstream/master' into BUG-regression-datetimeindex-slicing

simonjayhawkins commented:
@CloseChoice can you fix the pre-commit errors. (I'm on Windows and pre-commit fails to initialize and these checks can no longer be performed locally from code-checks.sh #37240)  cc @MarcoGorelli

MarcoGorelli commented:
> @CloseChoice can you fix the pre-commit errors. (I'm on Windows and pre-commit fails to initialize and these checks can no longer be performed locally from code-checks.sh #37240) cc @MarcoGorelli

@CloseChoice you just need to replace `pd.DataFrame` with `DataFrame`

@simonjayhawkins what errors do you get? Part of the motivation for moving these checks over to pre-commit was that they'd be cross-platform

MarcoGorelli commented on pandas/tests/indexing/test_partial.py:
```diff
@@ -681,3 +681,24 @@ def test_index_name_empty(self):
+    def test_slice_irregular_datetime_index_with_nan(self):
+        # GH36953
+        index = pd.to_datetime(["2012-01-01", "2012-01-02", "2012-01-03", None])
+        df = pd.DataFrame(range(len(index)), index=index)
```
```suggestion
        df = DataFrame(range(len(index)), index=index)
```

MarcoGorelli commented on pandas/tests/indexing/test_partial.py:
```diff
@@ -681,3 +681,24 @@ def test_index_name_empty(self):
+        # GH36953
+        index = pd.to_datetime(["2012-01-01", "2012-01-02", "2012-01-03", None])
+        df = pd.DataFrame(range(len(index)), index=index)
+        expected = pd.DataFrame(range(len(index[:3])), index=index[:3])
```
```suggestion
        expected = DataFrame(range(len(index[:3])), index=index[:3])
```

MarcoGorelli commented on pandas/tests/indexing/test_partial.py:
```diff
@@ -681,3 +681,24 @@ def test_index_name_empty(self):
+
+    def test_slice_datetime_index(self):
+        # GH35509
+        df = pd.DataFrame(
```
```suggestion
        df = DataFrame(
```

MarcoGorelli commented on pandas/tests/indexing/test_partial.py:
```diff
@@ -681,3 +681,24 @@ def test_index_name_empty(self):
+            {"col1": ["a", "b", "c"], "col2": [1, 2, 3]},
+            index=pd.to_datetime(["2020-08-01", "2020-07-02", "2020-08-05"]),
+        )
+        expected = pd.DataFrame(
```
```suggestion
        expected = DataFrame(
```

MarcoGorelli committed:
Update pandas/tests/indexing/test_partial.py

MarcoGorelli committed:
Update pandas/tests/indexing/test_partial.py

MarcoGorelli committed:
Update pandas/tests/indexing/test_partial.py

MarcoGorelli committed:
Update pandas/tests/indexing/test_partial.py

simonjayhawkins commented:
> @simonjayhawkins what errors do you get? 
i'm using a different machine, so have not looked in detail yet. (and can't remember what did last time)

<details>

### version information

```
pre-commit version: 2.7.1
sys.version:
    3.8.6 | packaged by conda-forge | (default, Oct  7 2020, 18:22:52) [MSC v.1916 64 bit (AMD64)]
sys.executable: C:\Users\simon\anaconda3\envs\pandas-dev\python.exe
os.name: nt
sys.platform: win32
```

### error information

```
An unexpected error has occurred: CalledProcessError: command: ('C:\\Users\\simon\\anaconda3\\envs\\pandas-dev\\python.exe', '-mvirtualenv', 'C:\\Users\\simon\\.cache\\pre-commit\\repoyew02fci\\py_env-python3', '-p', 'C:\\Users\\simon\\anaconda3\\envs\\pandas-dev\\python.exe')
return code: 1
expected return code: 0
stdout:
    FileNotFoundError: [Errno 2] No such file or directory: 'C:\\Users\\simon\\anaconda3\\envs\\pandas-dev\\Lib\\venv\\scripts\\nt\\python.exe'
    
stderr: (none)
```

```
Traceback (most recent call last):
  File "C:\Users\simon\anaconda3\envs\pandas-dev\lib\site-packages\pre_commit\error_handler.py", line 63, in error_handler
    yield
  File "C:\Users\simon\anaconda3\envs\pandas-dev\lib\site-packages\pre_commit\main.py", line 390, in main
    return run(args.config, store, args)
  File "C:\Users\simon\anaconda3\envs\pandas-dev\lib\site-packages\pre_commit\commands\run.py", line 388, in run
    install_hook_envs(hooks, store)
  File "C:\Users\simon\anaconda3\envs\pandas-dev\lib\site-packages\pre_commit\repository.py", line 206, in install_hook_envs
    _hook_install(hook)
  File "C:\Users\simon\anaconda3\envs\pandas-dev\lib\site-packages\pre_commit\repository.py", line 82, in _hook_install
    lang.install_environment(
  File "C:\Users\simon\anaconda3\envs\pandas-dev\lib\site-packages\pre_commit\languages\python.py", line 213, in install_environment
    cmd_output_b(*venv_cmd, cwd='/')
  File "C:\Users\simon\anaconda3\envs\pandas-dev\lib\site-packages\pre_commit\util.py", line 157, in cmd_output_b
    raise CalledProcessError(returncode, cmd, retcode, stdout_b, stderr_b)
pre_commit.util.CalledProcessError: command: ('C:\\Users\\simon\\anaconda3\\envs\\pandas-dev\\python.exe', '-mvirtualenv', 'C:\\Users\\simon\\.cache\\pre-commit\\repoyew02fci\\py_env-python3', '-p', 'C:\\Users\\simon\\anaconda3\\envs\\pandas-dev\\python.exe')
return code: 1
expected return code: 0
stdout:
    FileNotFoundError: [Errno 2] No such file or directory: 'C:\\Users\\simon\\anaconda3\\envs\\pandas-dev\\Lib\\venv\\scripts\\nt\\python.exe'
    
stderr: (none)

```

</details>

> Part of the motivation for moving these checks over to pre-commit was that they'd be cross-platform

I understand that, but didn't we discuss that pre-commit was optional.


> @simonjayhawkins the pre-commit hooks all pass, it's the code-checks.sh that fail 🤣

my bad. missed that.

MarcoGorelli commented:
@simonjayhawkins thanks for the traceback - are you using conda? I could reproduce that on a windows conda environment. Looks like it's a bug in virtualenv https://github.com/pypa/virtualenv/issues/1986

If you downgrade virtualenv to 20.0.33 then it should work (did for me) - if it works for you I'll note it in the docs

simonjayhawkins commented:
> If you downgrade virtualenv to 20.0.33 then it should work (did for me) - if it works for you I'll note it in the docs

thanks works for me too. (20.0.35 was installed). just started running on all-files. will report back if fails.

simonjayhawkins commented:
back to the issue

> so pls merge master and let's see what the CI says

failures on Windows py38_np18 seen on other PRs.

jreback commented:
thanks @CloseChoice

MarcoGorelli commented:
> will report back if fails.

although it installed fine for me, some hooks still failed, likely because I have e.g.
```
./scripts/validate_unwanted_patterns --validation-type="private_function_across_module"
```
instead of
```
python scripts/validate_unwanted_patterns --validation-type="private_function_across_module"
```
, will fix

simonjayhawkins commented:
> > will report back if fails.
> 
> although it installed fine for me, some hooks still failed, likely because currently we have

works for me in anaconda prompt and git bash
